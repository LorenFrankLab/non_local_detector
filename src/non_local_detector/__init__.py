from non_local_detector.models import (
    ContFragClusterlessClassifier,  # noqa
    ContFragSortedSpikesClassifier,  # noqa
    MultiEnvironmentSortedSpikesClassifier,  # noqa
    MultiEnvironmentClusterlessClassifier,  # noqa
    NonLocalClusterlessDetector,  # noqa
    NonLocalSortedSpikesDetector,  # noqa
    ClusterlessDecoder,  # noqa
    SortedSpikesDecoder,  # noqa
)

try:
    from ._version import __version__
except ImportError:
    pass
