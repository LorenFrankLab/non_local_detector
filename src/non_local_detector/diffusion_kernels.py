from typing import Any

import jax
import jax.numpy as jnp
import jax.scipy.linalg
import networkx as nx
import numpy as np


def _assign_gaussian_weights_from_distance(G: nx.Graph, bandwidth_sigma: float) -> None:
    """
    Overwrites each edge's "weight" attribute with
        w_uv = exp( - (distance_uv)^2 / (2 * sigma^2) ).
    Assumes each edge already has "distance" = Euclidean length.
    """
    two_sigma2 = 2.0 * (bandwidth_sigma**2)
    for u, v, data in G.edges(data=True):
        d = data.get("distance", None)
        if d is None:
            raise KeyError(f"Edge ({u},{v}) has no 'distance' attribute.")
        # compute the RBF weight
        data["weight"] = float(np.exp(-(d * d) / two_sigma2))


def compute_diffusion_kernels(
    graph: nx.Graph,
    bandwidth_sigma: float,
) -> jnp.ndarray:
    """
    Computes diffusion kernels for all valid bins using the
    Graph Laplacian method and matrix exponential.

    Parameters
    ----------
    graph : nx.Graph
        The graph representing the track, where nodes are bins and edges are connections.
    bandwidth_sigma : float
        The bandwidth of the Gaussian kernel. This controls the spread of the kernel.

    Returns
    -------
    kernel_matrix : jnp.ndarray, shape (n_bins, n_bins)
        The diffusion kernel matrix for the bins.
        Therepresents the amount of concentration (or probability mass, if u represents probability)
        that has flowed from node j to node i after time t.
    """

    n_bins = graph.number_of_nodes()
    _assign_gaussian_weights_from_distance(graph, bandwidth_sigma)
    laplacian_full = jnp.array(
        nx.laplacian_matrix(graph, nodelist=range(n_bins), weight="weight").toarray(),
        dtype=jnp.float32,
    )

    exponent_coefficient = bandwidth_sigma**2 / 2.0
    kernel_matrix = jax.scipy.linalg.expm(-exponent_coefficient * laplacian_full)

    # Clip to non-negative values
    kernel_matrix = jnp.clip(kernel_matrix, a_min=0.0, a_max=None)

    # Normalize the kernel matrix so the sum of each column is 1.
    col_sums = kernel_matrix.sum(axis=0, keepdims=True)
    kernel_matrix = jnp.where(col_sums == 0.0, 0.0, kernel_matrix / col_sums)

    return kernel_matrix
