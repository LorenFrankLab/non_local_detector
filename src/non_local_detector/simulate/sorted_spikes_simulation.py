from typing import Tuple

import numpy as np
from scipy.stats import multivariate_normal, norm

TRACK_HEIGHT = 170
SAMPLING_FREQUENCY = 1500


def simulate_poisson_spikes(rate, sampling_frequency):
    """Given a rate, returns a time series of spikes.
    Parameters
    ----------
    rate : np.ndarray, shape (n_time,)
    sampling_frequency : float
    Returns
    -------
    spikes : np.ndarray, shape (n_time,)
    """
    return np.random.poisson(rate / sampling_frequency)


def simulate_time(n_samples, sampling_frequency):
    return np.arange(n_samples) / sampling_frequency


def simulate_position(time, track_height, running_speed=10):
    half_height = track_height / 2
    return (
        half_height * np.sin(2 * np.pi * time / running_speed - np.pi / 2) + half_height
    )


def create_place_field(
    place_field_mean,
    position,
    sampling_frequency,
    is_condition=None,
    place_field_std_deviation=12.5,
    max_firing_rate=20,
    baseline_firing_rate=0.001,
):
    if is_condition is None:
        is_condition = np.ones_like(position, dtype=bool)
    field_firing_rate = norm(place_field_mean, place_field_std_deviation).pdf(position)
    field_firing_rate /= np.nanmax(field_firing_rate)
    field_firing_rate[~is_condition] = 0
    return baseline_firing_rate + max_firing_rate * field_firing_rate


def simulate_place_field_firing_rate(
    means, position, max_rate=15, variance=10, is_condition=None
):
    """Simulates the firing rate of a neuron with a place field at `means`.

    Parameters
    ----------
    means : ndarray, shape (n_position_dims,)
    position : ndarray, shape (n_time, n_position_dims)
    max_rate : float, optional
    variance : float, optional
    is_condition : None or ndarray, (n_time,)

    Returns
    -------
    firing_rate : ndarray, shape (n_time,)

    """
    if is_condition is None:
        is_condition = np.ones(position.shape[0], dtype=bool)
    firing_rate = multivariate_normal(means, variance).pdf(position)
    firing_rate /= firing_rate.max()
    firing_rate *= max_rate
    firing_rate[~is_condition] = 0.0

    return firing_rate


def simulate_neuron_with_place_field(
    means, position, max_rate=15, variance=36, sampling_frequency=500, is_condition=None
):
    """Simulates the spiking of a neuron with a place field at `means`.

    Parameters
    ----------
    means : ndarray, shape (n_position_dims,)
    position : ndarray, shape (n_time, n_position_dims)
    max_rate : float, optional
    variance : float, optional
    sampling_frequency : float, optional
    is_condition : None or ndarray, (n_time,)

    Returns
    -------
    spikes : ndarray, shape (n_time,)

    """
    firing_rate = simulate_place_field_firing_rate(
        means, position, max_rate, variance, is_condition
    )
    return simulate_poisson_spikes(firing_rate, sampling_frequency)


def get_trajectory_direction(position):
    is_inbound = np.insert(np.diff(position) < 0, 0, False)
    return np.where(is_inbound, "Inbound", "Outbound"), is_inbound


def gaussian_pdf(x, mean, sigma):
    """Compute the value of a Gaussian probability density function at x with
    given mean and sigma."""
    return np.exp(-0.5 * ((x - mean) / sigma) ** 2) / (sigma * np.sqrt(2.0 * np.pi))


def estimate_position_distance(
    place_bin_centers: np.ndarray,
    positions: np.ndarray,
    position_std: np.ndarray,
) -> np.ndarray:
    """Estimates the Euclidean distance between positions and position bins.

    Parameters
    ----------
    place_bin_centers : np.ndarray, shape (n_position_bins, n_position_dims)
    positions : np.ndarray, shape (n_time, n_position_dims)
    position_std : array_like, shape (n_position_dims,)

    Returns
    -------
    position_distance : np.ndarray, shape (n_time, n_position_bins)

    """
    n_time, n_position_dims = positions.shape
    n_position_bins = place_bin_centers.shape[0]

    position_distance = np.ones((n_time, n_position_bins), dtype=np.float32)

    if isinstance(position_std, (int, float)):
        position_std = [position_std] * n_position_dims

    for position_ind, std in enumerate(position_std):
        position_distance *= gaussian_pdf(
            np.expand_dims(place_bin_centers[:, position_ind], axis=0),
            np.expand_dims(positions[:, position_ind], axis=1),
            std,
        )

    return position_distance


def estimate_position_density(
    place_bin_centers: np.ndarray,
    positions: np.ndarray,
    position_std: np.ndarray,
    block_size: int = 100,
    sample_weights: np.ndarray = None,
) -> np.ndarray:
    """Estimates a kernel density estimate over position bins using
    Euclidean distances.

    Parameters
    ----------
    place_bin_centers : np.ndarray, shape (n_position_bins, n_position_dims)
    positions : np.ndarray, shape (n_time, n_position_dims)
    position_std : float or array_like, shape (n_position_dims,)
    sample_weights : None or np.ndarray, shape (n_time,)

    Returns
    -------
    position_density : np.ndarray, shape (n_position_bins,)

    """
    n_position_bins = place_bin_centers.shape[0]

    if block_size is None:
        block_size = n_position_bins

    position_density = np.empty((n_position_bins,))
    for start_ind in range(0, n_position_bins, block_size):
        block_inds = slice(start_ind, start_ind + block_size)
        position_density[block_inds] = np.average(
            estimate_position_distance(
                place_bin_centers[block_inds], positions, position_std
            ),
            axis=0,
            weights=sample_weights,
        )
    return position_density


def get_firing_rate(
    is_spike: np.ndarray,
    position: np.ndarray,
    place_bin_centers: np.ndarray,
    is_track_interior: np.ndarray,
    not_nan_position: np.ndarray,
    occupancy: np.ndarray,
    position_std: np.ndarray,
    block_size: int = None,
    weights: np.ndarray = None,
) -> np.ndarray:
    if is_spike.sum() > 0:
        mean_rate = np.average(is_spike, weights=weights)
        marginal_density = np.zeros((place_bin_centers.shape[0],), dtype=np.float32)

        marginal_density[is_track_interior] = estimate_position_density(
            place_bin_centers[is_track_interior],
            np.asarray(
                position[is_spike.astype(bool) & not_nan_position], dtype=np.float32
            ),
            position_std,
            block_size=block_size,
            sample_weights=np.asarray(
                weights[is_spike.astype(bool) & not_nan_position], dtype=np.float32
            ),
        )
        return np.spacing(1) + (mean_rate * marginal_density / occupancy)
    else:
        return np.zeros_like(occupancy)


def simulate_two_state_inhomogenous_poisson():
    track_height = 170
    sampling_frequency = 500
    n_samples = sampling_frequency * 240

    time = simulate_time(n_samples, sampling_frequency)
    position = simulate_position(time, track_height)[:, np.newaxis]

    _, is_inbound = get_trajectory_direction(position.squeeze())
    spikes = simulate_neuron_with_place_field(
        [50],
        position,
        max_rate=30,
        variance=36,
        sampling_frequency=sampling_frequency,
        is_condition=is_inbound,
    ) + simulate_neuron_with_place_field(
        [150],
        position,
        max_rate=10,
        variance=36,
        sampling_frequency=sampling_frequency,
        is_condition=~is_inbound,
    )

    return time, position, spikes


def simulate_linear_distance_with_pauses(
    time, track_height, running_speed=10, pause=0.5, sampling_frequency=1
):
    linear_distance = simulate_position(time, track_height, running_speed)
    peaks = np.nonzero(linear_distance == track_height)[0]
    n_pause_samples = int(pause * sampling_frequency)
    pause_linear_distance = np.zeros((time.size + n_pause_samples * peaks.size,))
    pause_ind = peaks[:, np.newaxis] + np.arange(n_pause_samples)
    pause_ind += np.arange(peaks.size)[:, np.newaxis] * n_pause_samples

    pause_linear_distance[pause_ind.ravel()] = track_height
    pause_linear_distance[pause_linear_distance == 0] = linear_distance
    return pause_linear_distance[: time.size]


def make_continuous_replay(
    n_neurons: int = 8,
    n_samples_between_spikes: int = 20,
    is_outbound: bool = True,
) -> np.ndarray:
    """Make a simulated continuous replay.

    Parameters
    ----------
    sampling_frequency : int, optional
        Samples per second
    track_height : float, optional
        Height of the simulated track
    running_speed : float, optional
        Speed of the simulated animal
    place_field_means : np.ndarray, optional
        Location of the center of the Gaussian place fields.
    replay_speedup : int, optional
        _description_, by default REPLAY_SPEEDUP
    is_outbound : bool, optional
        _description_, by default True

    Returns
    -------
    test_spikes : np.ndarray, shape (n_time, n_neurons)
        Binned spike indicator. 1 means spike occured. 0 means no spike occured.

    """
    neuron_order = (
        np.flip(np.arange(n_neurons), axis=0) if is_outbound else np.arange(n_neurons)
    )
    spike_time_ind = np.arange(
        0, n_neurons * n_samples_between_spikes, n_samples_between_spikes
    )
    spikes = np.zeros((spike_time_ind.max() + 1, n_neurons))
    spikes[(spike_time_ind, neuron_order)] = 1

    return spikes


def make_fragmented_replay(
    n_neurons: int = 8,
) -> Tuple[np.ndarray, np.ndarray]:
    """Make a simulated fragmented replay.

    Parameters
    ----------
    place_field_means : np.ndarray, optional
        _description_, by default PLACE_FIELD_MEANS
    sampling_frequency : int, optional
        Samples per second

    Returns
    -------
    test_spikes : np.ndarray, shape (n_time, n_neurons)
        Binned spike indicator. 1 means spike occured. 0 means no spike occured.

    """
    spike_time_ind = np.array([1, 21, 31, 41, 51, 61, 71, 81, 91, 101, 111, 121])
    neuron_ind = np.array([0, 5, 1, 5, 6, 5, 3, 0, 5, 1, 5, 6])
    test_spikes = np.zeros((spike_time_ind.max() + 1, n_neurons))
    test_spikes[(spike_time_ind, neuron_ind)] = 1.0

    return test_spikes


def make_simulated_data(
    track_height: float = TRACK_HEIGHT, sampling_frequency: int = SAMPLING_FREQUENCY
):
    n_samples = sampling_frequency * 65

    time = simulate_time(n_samples, sampling_frequency)
    linear_distance = (
        simulate_linear_distance_with_pauses(
            time, track_height, sampling_frequency=sampling_frequency, pause=3
        )
        + np.random.randn(*time.shape) * 1e-4
    )

    speed = np.abs(np.diff(linear_distance) / np.diff(time))
    speed = np.insert(speed, 0, 0.0)

    pause_ind = (
        np.nonzero(np.diff(np.isclose(linear_distance, track_height, atol=1e-5)))[0] + 1
    )

    pause_times = time[np.reshape(pause_ind, (-1, 2))]
    pause_width = np.diff(pause_times)[0]

    ripple_duration = 0.100

    mid_ripple_time = pause_times[:3, 0] + pause_width / 2
    ripple_times = (
        mid_ripple_time + np.array([-0.5, 0.5])[:, np.newaxis] * ripple_duration
    ).T

    place_field_means = np.arange(0, 200, 25)

    place_fields = np.stack(
        [
            create_place_field(place_field_mean, linear_distance, sampling_frequency)
            for place_field_mean in place_field_means
        ]
    )

    spikes = simulate_poisson_spikes(place_fields, sampling_frequency).T

    # Add replays

    # Ripple 1
    start_time, end_time = ripple_times[0]
    is_ripple_time = (time >= start_time) & (time <= end_time)
    ripple_ind = np.nonzero(is_ripple_time)[0]
    spikes[is_ripple_time] = 0
    ripple1_spikes = make_continuous_replay(
        n_neurons=spikes.shape[1],
        n_samples_between_spikes=20,
        is_outbound=True,
    )
    spikes[ripple_ind[: ripple1_spikes.shape[0]]] = ripple1_spikes

    # Ripple 2
    start_time, end_time = ripple_times[1]
    is_ripple_time = (time >= start_time) & (time <= end_time)
    ripple_ind = np.nonzero(is_ripple_time)[0]
    spikes[is_ripple_time] = 0
    ripple2_spikes = make_fragmented_replay(
        n_neurons=spikes.shape[1],
    )
    spikes[ripple_ind[: ripple2_spikes.shape[0]]] = ripple2_spikes

    # Ripple 3
    start_time, end_time = ripple_times[2]
    is_ripple_time = (time >= start_time) & (time <= end_time)
    ripple_ind = np.nonzero(is_ripple_time)[0]
    spikes[is_ripple_time] = 0
    ripple3_spikes = make_continuous_replay(
        n_neurons=spikes.shape[1],
        n_samples_between_spikes=20,
        is_outbound=False,
    )
    spikes[ripple_ind[: ripple3_spikes.shape[0]]] = ripple3_spikes

    # No spike scenario
    spikes[(time > 45) & (time < 47)] = 0.0

    return (speed, linear_distance, spikes, time, ripple_times, sampling_frequency)
