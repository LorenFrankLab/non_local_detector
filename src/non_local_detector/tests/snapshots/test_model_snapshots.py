"""Snapshot tests for decoder models using simulated data.

These tests use the syrupy snapshot testing framework to detect unintended
changes in model behavior. Snapshots capture expected outputs for regression
testing during refactoring and development.

The tests are based on the simulation framework from test_simulation.ipynb.
"""

import numpy as np
import pytest
from syrupy.assertion import SnapshotAssertion

from non_local_detector import (
    ContFragSortedSpikesClassifier,
    NonLocalSortedSpikesDetector,
    SortedSpikesDecoder,
)
from non_local_detector.simulate.sorted_spikes_simulation import make_simulated_data


@pytest.fixture
def simulated_data():
    """Generate simulated data with fixed random seed for reproducibility."""
    np.random.seed(42)  # Fixed seed for reproducibility
    (
        speed,
        position,
        spike_times,
        time,
        event_times,
        sampling_frequency,
        is_event,
        place_fields,
    ) = make_simulated_data(n_neurons=30)
    return {
        "speed": speed,
        "position": position,
        "spike_times": spike_times,
        "time": time,
        "event_times": event_times,
        "sampling_frequency": sampling_frequency,
        "is_event": is_event,
        "place_fields": place_fields,
    }


def serialize_xarray_summary(data_array):
    """Serialize xarray DataArray to summary statistics for snapshot comparison.

    For large arrays, we snapshot summary statistics rather than full data
    to keep snapshot files manageable and readable.
    """
    arr = np.asarray(data_array)
    return {
        "shape": arr.shape,
        "dtype": str(arr.dtype),
        "mean": float(np.mean(arr)),
        "std": float(np.std(arr)),
        "min": float(np.min(arr)),
        "max": float(np.max(arr)),
        "sum": float(np.sum(arr)),
        # Sample a few values for additional verification
        "first_5": arr.ravel()[:5].tolist() if arr.size >= 5 else arr.ravel().tolist(),
        "last_5": arr.ravel()[-5:].tolist() if arr.size >= 5 else arr.ravel().tolist(),
    }


def serialize_state_probabilities(state_probs):
    """Serialize state probabilities for snapshot comparison."""
    return {
        "shape": state_probs.shape,
        "states": list(state_probs.states.values),
        "mean_per_state": {
            str(state): float(np.mean(state_probs.sel(states=state)))
            for state in state_probs.states.values
        },
        "max_per_state": {
            str(state): float(np.max(state_probs.sel(states=state)))
            for state in state_probs.states.values
        },
        "min_per_state": {
            str(state): float(np.min(state_probs.sel(states=state)))
            for state in state_probs.states.values
        },
    }


@pytest.mark.snapshot
def test_nonlocal_sorted_spikes_detector_snapshot(
    simulated_data: dict, snapshot: SnapshotAssertion
):
    """Snapshot test for NonLocalSortedSpikesDetector.

    This test verifies that the NonLocalSortedSpikesDetector produces
    consistent outputs across code changes. It tests:
    - State probability predictions
    - Posterior distributions (summary statistics)
    - Model fitting results
    """
    detector = NonLocalSortedSpikesDetector(
        sorted_spikes_algorithm="sorted_spikes_kde",
        sorted_spikes_algorithm_params={
            "position_std": 6.0,
            "block_size": int(2**12),
        },
    ).fit(
        simulated_data["time"],
        simulated_data["position"],
        simulated_data["spike_times"],
        is_training=~simulated_data["is_event"],
    )

    results = detector.predict(
        spike_times=simulated_data["spike_times"],
        time=simulated_data["time"],
        position=simulated_data["position"],
        position_time=simulated_data["time"],
        save_log_likelihood_to_results=True,
    )

    # Snapshot state probabilities
    state_probs_summary = serialize_state_probabilities(
        results.acausal_state_probabilities
    )
    assert state_probs_summary == snapshot(name="state_probabilities")

    # Snapshot posterior summary (full posterior is too large)
    posterior_summary = serialize_xarray_summary(results.acausal_posterior)
    assert posterior_summary == snapshot(name="posterior_summary")

    # Snapshot model attributes
    model_summary = {
        "n_position_bins": detector.environments[0].place_bin_centers_.shape[0],
        "discrete_state_transitions_shape": detector.discrete_state_transitions_.shape,
        "discrete_state_transitions_sum_axis1": detector.discrete_state_transitions_.sum(
            axis=1
        ).tolist(),
        "initial_conditions_sum": float(detector.initial_conditions_.sum()),
        "initial_conditions_shape": detector.initial_conditions_.shape,
    }
    assert model_summary == snapshot(name="model_summary")


@pytest.mark.snapshot
def test_contfrag_sorted_spikes_classifier_snapshot(
    simulated_data: dict, snapshot: SnapshotAssertion
):
    """Snapshot test for ContFragSortedSpikesClassifier.

    This test verifies that the ContFragSortedSpikesClassifier produces
    consistent continuous vs. fragmented classification results.
    """
    classifier = ContFragSortedSpikesClassifier(
        sorted_spikes_algorithm="sorted_spikes_kde",
        sorted_spikes_algorithm_params={
            "position_std": 6.0,
            "block_size": int(2**12),
        },
    ).fit(
        position_time=simulated_data["time"],
        position=simulated_data["position"],
        spike_times=simulated_data["spike_times"],
        is_training=~simulated_data["is_event"],
    )

    results = classifier.predict(
        spike_times=simulated_data["spike_times"],
        time=simulated_data["time"],
    )

    # Snapshot state probabilities
    state_probs_summary = serialize_state_probabilities(
        results.acausal_state_probabilities
    )
    assert state_probs_summary == snapshot(name="state_probabilities")

    # Snapshot posterior summary
    posterior_summary = serialize_xarray_summary(results.acausal_posterior)
    assert posterior_summary == snapshot(name="posterior_summary")

    # Snapshot continuous state transitions (should sum to 2 along axis 1)
    transitions_summary = {
        "shape": classifier.continuous_state_transitions_.shape,
        "sum_axis1_mean": float(
            np.mean(classifier.continuous_state_transitions_.sum(axis=1))
        ),
        "sum_axis1_std": float(
            np.std(classifier.continuous_state_transitions_.sum(axis=1))
        ),
        "sum_axis1_min": float(
            np.min(classifier.continuous_state_transitions_.sum(axis=1))
        ),
        "sum_axis1_max": float(
            np.max(classifier.continuous_state_transitions_.sum(axis=1))
        ),
    }
    assert transitions_summary == snapshot(name="transitions_summary")


@pytest.mark.snapshot
def test_sorted_spikes_decoder_snapshot(
    simulated_data: dict, snapshot: SnapshotAssertion
):
    """Snapshot test for SortedSpikesDecoder.

    This test verifies that the basic SortedSpikesDecoder produces
    consistent position decoding results.
    """
    decoder = SortedSpikesDecoder(
        sorted_spikes_algorithm="sorted_spikes_kde",
        sorted_spikes_algorithm_params={
            "position_std": 6.0,
            "block_size": int(2**12),
        },
    ).fit(
        position_time=simulated_data["time"],
        position=simulated_data["position"],
        spike_times=simulated_data["spike_times"],
        is_training=~simulated_data["is_event"],
    )

    results = decoder.predict(
        spike_times=simulated_data["spike_times"],
        time=simulated_data["time"],
        position=simulated_data["position"],
        position_time=simulated_data["time"],
        save_log_likelihood_to_results=True,
    )

    # Snapshot state probabilities (should be high for continuous state)
    state_probs_summary = {
        "shape": results.acausal_state_probabilities.shape,
        "mean": float(np.mean(results.acausal_state_probabilities)),
        "min": float(np.min(results.acausal_state_probabilities)),
        "max": float(np.max(results.acausal_state_probabilities)),
    }
    assert state_probs_summary == snapshot(name="state_probabilities")

    # Snapshot posterior summary
    posterior_summary = serialize_xarray_summary(results.acausal_posterior)
    assert posterior_summary == snapshot(name="posterior_summary")


@pytest.mark.snapshot
@pytest.mark.parametrize(
    "algorithm,algorithm_params",
    [
        ("sorted_spikes_kde", {"position_std": 6.0, "block_size": int(2**12)}),
    ],
)
def test_nonlocal_detector_different_algorithms_snapshot(
    simulated_data: dict,
    algorithm: str,
    algorithm_params: dict,
    snapshot: SnapshotAssertion,
):
    """Snapshot test for NonLocalSortedSpikesDetector with different algorithms.

    This test verifies consistent behavior across different likelihood algorithms.
    Note: GLM algorithm is excluded as it requires additional environment parameters.
    """
    detector = NonLocalSortedSpikesDetector(
        sorted_spikes_algorithm=algorithm,
        sorted_spikes_algorithm_params=algorithm_params,
    ).fit(
        simulated_data["time"],
        simulated_data["position"],
        simulated_data["spike_times"],
        is_training=~simulated_data["is_event"],
    )

    results = detector.predict(
        spike_times=simulated_data["spike_times"],
        time=simulated_data["time"],
        position=simulated_data["position"],
        position_time=simulated_data["time"],
    )

    # Snapshot state probabilities
    state_probs_summary = serialize_state_probabilities(
        results.acausal_state_probabilities
    )
    assert state_probs_summary == snapshot


@pytest.mark.snapshot
def test_detector_encoding_model_snapshot(
    simulated_data: dict, snapshot: SnapshotAssertion
):
    """Snapshot test for encoding model properties.

    This test verifies that the fitted encoding models have expected properties.
    """
    detector = NonLocalSortedSpikesDetector(
        sorted_spikes_algorithm="sorted_spikes_kde",
        sorted_spikes_algorithm_params={
            "position_std": 6.0,
            "block_size": int(2**12),
        },
    ).fit(
        simulated_data["time"],
        simulated_data["position"],
        simulated_data["spike_times"],
        is_training=~simulated_data["is_event"],
    )

    # Get encoding model for first environment
    enc_model = detector.encoding_model_[("", 0)]

    # Snapshot place fields summary
    place_fields = enc_model["place_fields"]
    place_fields_summary = {
        "n_neurons": len(place_fields),
        "place_field_shape": place_fields[0].shape if len(place_fields) > 0 else None,
        "mean_rates": [float(np.mean(pf)) for pf in place_fields[:5]],  # First 5
        "max_rates": [float(np.max(pf)) for pf in place_fields[:5]],
        "argmax_positions": [int(np.argmax(pf)) for pf in place_fields[:5]],
    }
    assert place_fields_summary == snapshot(name="place_fields_summary")

    # Snapshot occupancy
    occupancy_summary = {
        "shape": enc_model["occupancy"].shape,
        "mean": float(np.mean(enc_model["occupancy"])),
        "min": float(np.min(enc_model["occupancy"])),
        "max": float(np.max(enc_model["occupancy"])),
        "sum": float(np.sum(enc_model["occupancy"])),
    }
    assert occupancy_summary == snapshot(name="occupancy_summary")
