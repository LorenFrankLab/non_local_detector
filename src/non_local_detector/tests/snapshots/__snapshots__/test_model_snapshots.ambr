# serializer version: 1
# name: test_contfrag_sorted_spikes_classifier_snapshot[posterior_summary]
  dict({
    'dtype': 'float32',
    'first_5': list([
      0.00803915224969387,
      0.008032179437577724,
      0.008021487854421139,
      0.008006716147065163,
      0.007988138124346733,
    ]),
    'last_5': list([
      0.003568395273759961,
      0.003570385044440627,
      0.003572101006284356,
      0.003573428140953183,
      0.003574422560632229,
    ]),
    'max': 0.2712382674217224,
    'mean': 0.00581395672634244,
    'min': 1.366967804441288e-31,
    'shape': tuple(
      97500,
      172,
    ),
    'std': 0.015416691079735756,
    'sum': 97500.0546875,
  })
# ---
# name: test_contfrag_sorted_spikes_classifier_snapshot[state_probabilities]
  dict({
    'max_per_state': dict({
      'Continuous': 0.9999502897262573,
      'Fragmented': 0.9754221439361572,
    }),
    'mean_per_state': dict({
      'Continuous': 0.7639068961143494,
      'Fragmented': 0.23609308898448944,
    }),
    'min_per_state': dict({
      'Continuous': 0.02457786537706852,
      'Fragmented': 4.973888280801475e-05,
    }),
    'shape': tuple(
      97500,
      2,
    ),
    'states': list([
      np.str_('Continuous'),
      np.str_('Fragmented'),
    ]),
  })
# ---
# name: test_contfrag_sorted_spikes_classifier_snapshot[transitions_summary]
  dict({
    'shape': tuple(
      172,
      172,
    ),
    'sum_axis1_max': 2.0000000000000004,
    'sum_axis1_mean': 2.0,
    'sum_axis1_min': 1.9999999999999996,
    'sum_axis1_std': 1.9154948419142388e-16,
  })
# ---
# name: test_detector_encoding_model_snapshot[occupancy_summary]
  dict({
    'max': 0.022265903651714325,
    'mean': 0.0047567603178322315,
    'min': 0.002888048067688942,
    'shape': tuple(
      86,
    ),
    'sum': 0.409081369638443,
  })
# ---
# name: test_detector_encoding_model_snapshot[place_fields_summary]
  dict({
    'argmax_positions': list([
      0,
      1,
      6,
      9,
      12,
    ]),
    'max_rates': list([
      0.009271292015910149,
      0.009650791995227337,
      0.006796794943511486,
      0.005789443384855986,
      0.006383868865668774,
    ]),
    'mean_rates': list([
      0.0005235845455899835,
      0.0008181324228644371,
      0.0006972372648306191,
      0.0005797083722427487,
      0.0006156781455501914,
    ]),
    'n_neurons': 30,
    'place_field_shape': tuple(
      86,
    ),
  })
# ---
# name: test_nonlocal_detector_different_algorithms_snapshot[sorted_spikes_kde-algorithm_params0]
  dict({
    'max_per_state': dict({
      'Local': 1.0,
      'No-Spike': 0.007492624688893557,
      'Non-Local Continuous': 0.9980527758598328,
      'Non-Local Fragmented': 0.9927268624305725,
    }),
    'mean_per_state': dict({
      'Local': 0.9751313924789429,
      'No-Spike': 0.00031108991242945194,
      'Non-Local Continuous': 0.01619352586567402,
      'Non-Local Fragmented': 0.008364012464880943,
    }),
    'min_per_state': dict({
      'Local': 3.2001970418085534e-30,
      'No-Spike': 0.0,
      'Non-Local Continuous': 0.0,
      'Non-Local Fragmented': 0.0,
    }),
    'shape': tuple(
      97500,
      4,
    ),
    'states': list([
      np.str_('Local'),
      np.str_('No-Spike'),
      np.str_('Non-Local Continuous'),
      np.str_('Non-Local Fragmented'),
    ]),
  })
# ---
# name: test_nonlocal_sorted_spikes_detector_snapshot[model_summary]
  dict({
    'discrete_state_transitions_shape': tuple(
      4,
      4,
    ),
    'discrete_state_transitions_sum_axis1': list([
      1.0,
      1.0,
      1.0,
      1.0,
    ]),
    'initial_conditions_shape': tuple(
      174,
    ),
    'initial_conditions_sum': 1.0,
    'n_position_bins': 86,
  })
# ---
# name: test_nonlocal_sorted_spikes_detector_snapshot[posterior_summary]
  dict({
    'dtype': 'float32',
    'first_5': list([
      1.0,
      0.0,
      0.0,
      0.0,
      0.0,
    ]),
    'last_5': list([
      5.739624975831248e-05,
      5.7428253057878464e-05,
      5.745585440308787e-05,
      5.7477198424749076e-05,
      5.749319461756386e-05,
    ]),
    'max': 1.0,
    'mean': 0.0057471212930977345,
    'min': 0.0,
    'shape': tuple(
      97500,
      174,
    ),
    'std': 0.07410204410552979,
    'sum': 97499.9140625,
  })
# ---
# name: test_nonlocal_sorted_spikes_detector_snapshot[state_probabilities]
  dict({
    'max_per_state': dict({
      'Local': 1.0,
      'No-Spike': 0.007492624688893557,
      'Non-Local Continuous': 0.9980527758598328,
      'Non-Local Fragmented': 0.9927268624305725,
    }),
    'mean_per_state': dict({
      'Local': 0.9751313924789429,
      'No-Spike': 0.00031108991242945194,
      'Non-Local Continuous': 0.01619352586567402,
      'Non-Local Fragmented': 0.008364012464880943,
    }),
    'min_per_state': dict({
      'Local': 3.2001970418085534e-30,
      'No-Spike': 0.0,
      'Non-Local Continuous': 0.0,
      'Non-Local Fragmented': 0.0,
    }),
    'shape': tuple(
      97500,
      4,
    ),
    'states': list([
      np.str_('Local'),
      np.str_('No-Spike'),
      np.str_('Non-Local Continuous'),
      np.str_('Non-Local Fragmented'),
    ]),
  })
# ---
# name: test_sorted_spikes_decoder_snapshot[posterior_summary]
  dict({
    'dtype': 'float32',
    'first_5': list([
      0.05160290002822876,
      0.051434893161058426,
      0.051135268062353134,
      0.05067235603928566,
      0.05004323273897171,
    ]),
    'last_5': list([
      7.546769470196529e-14,
      8.687820428343596e-14,
      9.456822667617756e-14,
      9.183136835590883e-14,
      7.031884568988594e-14,
    ]),
    'max': 0.2717386782169342,
    'mean': 0.011627902276813984,
    'min': 0.0,
    'shape': tuple(
      97500,
      86,
    ),
    'std': 0.027332890778779984,
    'sum': 97499.9609375,
  })
# ---
# name: test_sorted_spikes_decoder_snapshot[state_probabilities]
  dict({
    'max': 1.0000003576278687,
    'mean': 1.0,
    'min': 0.9999995827674866,
    'shape': tuple(
      97500,
    ),
  })
# ---
